Tracking "submission" events.

alex-mo opened the pull request.

alex-mo committed:
campaignion_tracking: README.

alex-mo committed:
campaignion_tracking: Update packages.

alex-mo committed:
campaignion_tracking: Pre-commit hooks.

alex-mo committed:
campaignion_tracking: Lint code.

alex-mo committed:
campaignion_tracking: "submission" event.

torotil commented on campaignion_tracking/campaignion_tracking.module:
```diff
@@ -112,12 +118,14 @@ function campaignion_tracking_webform_redirect_alter($redirect, $submission) {
 
-  $redirect->fragment = empty($redirect->fragment) ? $fragment : $redirect->fragment . ';' . $fragment;
+  $tracking_fragment = implode(';', $fragments);
+  $redirect->fragment = empty($redirect->fragment) ? $tracking_fragment : $redirect->fragment . ';' . $tracking_fragment;
```
How does this work wrt url encoding? Why are the values individually prefixed instead of using `&`-separated key value pairs?

alex-mo replied:
Is there a convention to use `&` in our fragment encoding?

alex-mo replied:
Can you give an example how you would suggest to encode the information in the fragment?

torotil replied:
The idea was to have something like this:

```
#real-fragment-without-namespace;namespace1:key1=value1&key2=value2;namespace2:key1=value1&…
```

alex-mo replied:
I see.

Although there is not much difference besides brevity to
```
#real-fragment-without-namespace;namespace1:key1=value1;namespace1:key2=value2;namespace2:key1=value1;…
```

- Without `&` parsing/encoding is simpler.
- Namespaces in this module are quite short (`t`, `d`, `w`), so not much added length
- Concepts stay the same

alex-mo replied:
We probably can just add splitting by `&` to the parser. If namespaces appear multiple times, they just get merged. This is currently how the tracking code work. Easier to just add additional information then...

torotil replied:
This scheme was invented with webform_prefill. I’d like to keep it consistent across projects. If the tracking code works differently then this is a bug.

torotil replied:
I don’t think parsing is simpler: This way parsing is.

1. split by `;`
2. Find parts that start with the right `prefix:`.
3. Use a parser for URL-parameters (usually in the standard library) to parse the string and get all key value pairs and decode them.

torotil reviewed (changes requested):
Apart from the question in the other comment:

- Yes, this needs indeed rebasing.
- ~~The recompile JS commit shouldn’t be included in the PR.~~
- Tests for the PHP-part are failing, they also might need additional coverage.

alex-mo commented:
- Rebased
- Removed compiled assets

alex-mo committed:
campaignion_tracking: Update babel

- fix issues with node 12.17+ see https://github.com/nodejs/node/issues/32852

alex-mo committed:
campaignion_tracking: Do not useBuiltIns (JS)

- Polyfill should be added by the themes

alex-mo committed:
campaignion_tracking: Refactor parsing of fragment

- the specification of how to encode the fragment changed
- the spec is now in line with webform_prefill

alex-mo committed:
campaignion_tracking: Run test in pre-commit hook

alex-mo committed:
campaignion_tracking: Refactor tracking fragment

alex-mo committed:
campaignion_tracking: Format file

alex-mo committed:
campaignion_tracking: Adapt test to new spec

alex-mo committed:
campaignion_tracking: no need for main.js

- main.js was only used for and included by the index.html for
  testing/prototyping purposes
- this can also be achieved with a <script type="module"> code block in
  index.html

torotil commented on campaignion_tracking/package/src/gtm.js:
```diff
@@ -154,11 +171,59 @@ export class GTMTracker {
+      }
+    }
+    // Allow others to modify the data being sent to GTM.
+    data = this.callChangeHook(eventName, data, this._context)
```
Is there a reason to implement this in hook-style as opposed to JS events?

alex-mo replied:
how would you receive back the modified data then? :thinking: 

but I agree, there might be opportunity for refactoring here...

torotil commented on campaignion_tracking/package/src/fragment.js:
```diff
@@ -57,35 +63,40 @@ export function consumeLocationHashForPrefixes (prefixes = [], locationHash = ''
+export function parsePartsWithPrefix (partsString, prefix = '') {
+  const split = partsString.split('&')
+  const parts = split.map(part => {
+    const _p = parsePart(part)
```
Does this handle URL-encoding? (I’m not sure where’s the best place to add the comment)

alex-mo replied:
Should be not an issue wrt `&`, they are not escaped:

```
> window.location.hash = '& '
"& "
> window.location.hash
"#&%20"
```

(for Fx and Chroumium)

torotil replied:
The parser still needs to handle those values properly and pass them *decoded* to the events.

How about `w:title=One: & two; & three,`?

alex-mo replied:
OK, thanks, got it...

alex-mo replied:
Should be fixed now.
- Assuming all values need to be encoded
- encoding only relevant string here: `title` (the other values are hardcoded chars or integers)

torotil commented on campaignion_tracking/campaignion_tracking.module:
```diff
@@ -119,13 +120,19 @@ function campaignion_tracking_webform_redirect_alter($redirect, $submission) {
+  $fragments[] = 'w:'. implode('&', $webformFragments);
+  if ($donationFragments) {
+    $fragments[] = 'd:'. implode('&', $donationFragments);
+  }
```
Why do we use 3 prefixes for a single module?

alex-mo replied:
this was the original design which I left untouched for the moment... idea was to have different prefixes for different "sources"...

happy to use just one `t` namespace as well if this is more appreciated :)

alex-mo replied:
Should we refactor this now? (I guess now is better than "later"...)

torotil replied:
How about `w:title=One: & two; & three,`?

torotil replied:
I’m not sure whether now is the best time. The original intention was to give each module (or use-case) one prefix so that they don’t interfere with each other.

torotil reviewed (commented):
Still a few questions … Nothing that I consider blockers though.

alex-mo committed:
campaignion_tracking: string encoding in fragment

alex-mo committed:
campaignion_tracking: Update bundled tracking JS
